### ISSUE TITLE
continue training of the deeplab-residual model, the loss fucntion will diverage.

### ISSUE THREAD

**Issue by gaopeng-eugene** [DrSleep/tensorflow-deeplab-resnet#5]
I use the PASCAL VOC 2012 AUG dataset, here is the loss 

step 500 	 loss = 0.221, (20.800 sec/step)
step 501 	 loss = 0.115, (0.445 sec/step)
step 502 	 loss = 0.115, (0.425 sec/step)
step 503 	 loss = 0.204, (0.420 sec/step)
step 504 	 loss = 0.104, (0.418 sec/step)
step 505 	 loss = 0.098, (0.418 sec/step)
step 506 	 loss = 0.219, (0.417 sec/step)
step 507 	 loss = 0.128, (0.419 sec/step)
step 508 	 loss = 0.129, (0.419 sec/step)
step 509 	 loss = 0.149, (0.421 sec/step)
step 510 	 loss = 0.117, (0.418 sec/step)
step 511 	 loss = 0.091, (0.489 sec/step)
step 512 	 loss = 0.341, (0.546 sec/step)
step 513 	 loss = 0.109, (0.517 sec/step)
step 514 	 loss = 0.176, (0.517 sec/step)
step 515 	 loss = 0.181, (0.559 sec/step)
step 516 	 loss = 0.140, (0.524 sec/step)
step 517 	 loss = 0.117, (0.538 sec/step)
step 518 	 loss = 0.111, (0.519 sec/step)
step 519 	 loss = 0.213, (0.516 sec/step)
step 520 	 loss = 0.192, (0.518 sec/step)
step 521 	 loss = 0.424, (0.509 sec/step)
step 522 	 loss = 0.096, (0.492 sec/step)
step 523 	 loss = 0.169, (0.537 sec/step)
step 524 	 loss = 0.171, (0.493 sec/step)
step 525 	 loss = 0.158, (0.536 sec/step)
step 526 	 loss = 0.150, (0.506 sec/step)
step 527 	 loss = 0.201, (0.549 sec/step)
step 528 	 loss = 0.146, (0.511 sec/step)
step 529 	 loss = 0.134, (0.507 sec/step)
step 530 	 loss = 0.162, (0.496 sec/step)
step 531 	 loss = 0.234, (0.524 sec/step)
step 532 	 loss = 0.228, (0.572 sec/step)
step 533 	 loss = 0.115, (0.529 sec/step)
step 534 	 loss = 0.174, (0.522 sec/step)
step 535 	 loss = 0.149, (0.504 sec/step)
step 536 	 loss = 0.206, (0.525 sec/step)
step 537 	 loss = 0.221, (0.542 sec/step)
step 538 	 loss = 0.093, (0.513 sec/step)
step 539 	 loss = 0.251, (0.565 sec/step)
step 540 	 loss = 0.084, (0.744 sec/step)
step 541 	 loss = 0.086, (0.534 sec/step)
step 542 	 loss = 0.175, (0.560 sec/step)
step 543 	 loss = 0.058, (0.538 sec/step)
step 544 	 loss = 0.214, (0.520 sec/step)
step 545 	 loss = 0.124, (0.498 sec/step)
step 546 	 loss = 0.097, (0.535 sec/step)
step 547 	 loss = 0.172, (0.547 sec/step)
step 548 	 loss = 0.234, (0.549 sec/step)
step 549 	 loss = 0.186, (0.518 sec/step)
step 550 	 loss = 0.262, (0.518 sec/step)
step 551 	 loss = 0.132, (0.522 sec/step)
step 552 	 loss = 0.156, (0.502 sec/step)
step 553 	 loss = 0.066, (0.534 sec/step)
step 554 	 loss = 0.155, (0.573 sec/step)
step 555 	 loss = 0.145, (0.530 sec/step)
step 556 	 loss = 0.225, (0.513 sec/step)
step 557 	 loss = 0.136, (0.519 sec/step)
step 558 	 loss = 0.223, (0.499 sec/step)
step 559 	 loss = 0.109, (0.532 sec/step)
step 560 	 loss = 0.133, (0.523 sec/step)
step 561 	 loss = 0.107, (0.534 sec/step)
step 562 	 loss = 0.247, (0.523 sec/step)
step 563 	 loss = 0.084, (0.528 sec/step)
step 564 	 loss = 0.172, (0.508 sec/step)
step 565 	 loss = 0.125, (0.504 sec/step)
step 566 	 loss = 0.244, (0.531 sec/step)
step 567 	 loss = 0.144, (0.536 sec/step)
step 568 	 loss = 0.110, (0.562 sec/step)
step 569 	 loss = 0.111, (0.513 sec/step)
step 570 	 loss = 6.620, (0.511 sec/step)
step 571 	 loss = 2.842, (0.539 sec/step)
step 572 	 loss = 5.740, (0.515 sec/step)
step 573 	 loss = 4.940, (0.513 sec/step)
step 574 	 loss = 4.994, (0.535 sec/step)
step 575 	 loss = 2.670, (0.522 sec/step)
step 576 	 loss = 2.900, (0.537 sec/step)
step 577 	 loss = 2.638, (0.513 sec/step)
step 578 	 loss = 3.622, (0.529 sec/step)
step 579 	 loss = 2.524, (0.531 sec/step)
step 580 	 loss = 2.554, (0.586 sec/step)
step 581 	 loss = 3.612, (0.525 sec/step)
step 582 	 loss = 2.171, (0.512 sec/step)
step 583 	 loss = 3.751, (0.548 sec/step)
step 584 	 loss = 2.245, (0.506 sec/step)
step 585 	 loss = 3.306, (0.536 sec/step)
step 586 	 loss = 1.847, (0.526 sec/step)
step 587 	 loss = 4.644, (0.556 sec/step)
step 588 	 loss = 2.619, (0.513 sec/step)
step 589 	 loss = 3.449, (0.534 sec/step)
step 590 	 loss = 2.106, (0.528 sec/step)
step 591 	 loss = 1.349, (0.519 sec/step)
step 592 	 loss = 1.875, (0.557 sec/step)
step 593 	 loss = 2.142, (0.527 sec/step)
step 594 	 loss = 1.511, (0.529 sec/step)
step 595 	 loss = 2.507, (0.537 sec/step)
step 596 	 loss = 3.175, (0.531 sec/step)
step 597 	 loss = 2.369, (0.536 sec/step)
step 598 	 loss = 1.895, (0.521 sec/step)
step 599 	 loss = 2.188, (0.533 sec/step)
The checkpoint has been created.
step 600 	 loss = 0.994, (21.880 sec/step)
step 601 	 loss = 2.987, (0.433 sec/step)
step 602 	 loss = 2.304, (0.417 sec/step)
step 603 	 loss = 2.919, (0.418 sec/step)
step 604 	 loss = 2.105, (0.416 sec/step)
step 605 	 loss = 1.149, (0.422 sec/step)
step 606 	 loss = 1.857, (0.416 sec/step)
step 607 	 loss = 2.461, (0.415 sec/step)
step 608 	 loss = 1.290, (0.415 sec/step)
step 609 	 loss = 1.688, (0.416 sec/step)
step 610 	 loss = 2.122, (0.418 sec/step)
step 611 	 loss = 2.046, (0.447 sec/step)
step 612 	 loss = 2.429, (0.523 sec/step)
step 613 	 loss = 2.406, (0.520 sec/step)

**Comment by gaopeng-eugene** [comment]
When I evaluate the diveraged model, the result is 0.034.

**Comment by gaopeng-eugene** [comment]
If I output the prediction results, its all black.

**Comment by DrSleep** [comment]
The divergence is likely happening due to the fact that the batch normalisation layer used in kaffe-tensorflow is only tested for inference. 
I have not noticed it before, so will update the code accordingly.

**Comment by gaopeng-eugene** [comment]
Have you ever trained the residual version or you only test the model ?

Thanks for your quick response.

**Comment by gaopeng-eugene** [comment]
If our target is to train the model from scratch, can we initialize the model from some tensor-flow residual network. By doing this, we can avoid the kaffe problem.

**Comment by DrSleep** [comment]
Yes, I have trained it, and also encountered the divergence after some number of iterations. As a temporary solution, I just decreased the learning rate, and it helped.

That is true, and I tried to use the model provided here: https://github.com/ry/tensorflow-resnet, but I wasn't able to produce nice results. I am not aware of other pre-trained ResNet models in TF.

As the DeepLab team also provides the init model of their network, I will try to re-train it with a correct batch norm implementation over the weekend.

**Comment by DrSleep** [comment]
I have created another branch with correct batch normalisation using the slim wrapper: https://github.com/DrSleep/tensorflow-deeplab-resnet/tree/batch-norm
The init model is being fine-tuned without any problems so far (please refer to the fine_tune.py script there), and gives around 60% mIoU after only 200 iterations. 

The init and pre-trained models are provided here: https://drive.google.com/drive/folders/0B_rootXHuswsZ3UtcWxZdHFBVzQ?usp=sharing

**Comment by gaopeng-eugene** [comment]
Thank you so much

**Comment by gaopeng-eugene** [comment]
Hi, can you provide the command line about how to doanload the latest version?

git clone can only download the old script.

**Comment by DrSleep** [comment]
Sure, you can run 
`git clone -b batch-norm https://github.com/DrSleep/tensorflow-deeplab-resnet.git`

**Comment by gaopeng-eugene** [comment]
Learn a lot from you. 

Do you have any plan on creating an multi-gpu version ? Batch normalization will be more powerful with larger batch size. And also, we need larger GPU memory for multi scale training which can produce better results.

**Comment by DrSleep** [comment]
Unfortunately, I do not have a multi-gpu setting at the moment, but yes, I agree with your points.
At the moment, freezing the mean and variance from the pre-trained models still produces tolerable results, but of course, for a new dataset that might not be the case anymore

**Comment by gaopeng-eugene** [comment]
what is the difference between deeplab_pretrained_bn.ckpt and deeplab_resnet.ckpt ? They can achieve the same evaluation result.

My guess is that there weights is the same while they use different tensorflow interface.

**Comment by DrSleep** [comment]
Yes, you are right. The 'bn' one is for the batch-norm branch, and another one is for the current master branch. The weights are the same, the names for the batch norm layer are different (as the slim interface prepends the parameters name with 'BatchNorm/').

**Comment by gaopeng-eugene** [comment]
Everything looks nice. I will let you know if there is other problems.

**Comment by gaopeng-eugene** [comment]
This one works so good. I finetuned the pretrained model on VOC2012 dataset for another 1000 times. The validation mAP raised to 0.789. This project will be very popular in semantic segmentation community.

Thank you so much for your great contribution.
